docs/spec_draft.md

```diff
@@ -1928,6 +1929,123 @@ produces an implementation-defined value.
 
 [Back to Ops](#index-of-ops)
 
+## stablehlo.dot_general
+
+### Semantics
+
+Computes dot products between slices of `lhs` and slices of `rhs` and produces a
+`result` tensor.
+
+More formally, `result[result_index] = dot_product`, where:
+  * `lhs_result_dimensions = [d for d in axes(lhs) and d not in lhs_batching_dimensions and d not in lhs_contracting_dimensions]`.
+  * `rhs_result_dimensions = [d for d in axes(rhs) and d not in rhs_batching_dimensions and d not in rhs_contracting_dimensions]`.
+  * `result_batching_index + result_lhs_index + result_rhs_index = result_index`
+    where `size(result_batching_index) = size(lhs_batching_dimensions)`,
+    `size(result_lhs_index) = size(lhs_result_dimensions)` and
+    `size(result_rhs_index) = size(rhs_result_dimensions)`.
+  * `transposed_lhs = transpose(lhs, lhs_batching_dimensions + lhs_result_dimensions + lhs_contracting_dimensions)`.
+  * `transposed_lhs_slice = slice(result_batching_index + result_lhs_index + [:, ...:])`.
+  * `reshaped_lhs_slice = reshape(transposed_lhs_slice, dims(lhs, lhs_contracting_dimensions))`.
+  * `transposed_rhs = transpose(rhs, rhs_batching_dimensions + rhs_result_dimensions + rhs_contracting_dimensions)`.
+  * `transposed_rhs_slice = slice(result_batching_index + result_rhs_index + [:, ...:])`.
+  * `reshaped_rhs_slice = reshape(transposed_rhs_slice, dims(rhs, rhs_contracting_dimensions))`.
+  * `dot_product = reduce(
+    inputs=[multiply(reshaped_lhs_slice, reshaped_rhs_slice)],
+    init_values=[0],
+    dimensions=[0, ..., size(lhs_contracting_dimensions) - 1],
+    body=lambda x, y: add(x, y))`.
+
+`precision_config` controls the tradeoff between speed and accuracy for
+computations on accelerator backends. This can be one of the following:
+  * `DEFAULT`: Fastest calculation, but least accurate approximation to the
+    original number.
+  * `HIGH`: Slower calculation, but more accurate approximation to the
+    original number.
+  * `HIGHEST`: Slowest calculation, but most accurate approximation to the
+    original number.
+
+### Inputs
+
+| Name                         | Type                                                        |
+|------------------------------|-------------------------------------------------------------|
+| `lhs`                        | tensor of any supported type                                |
+| `rhs`                        | tensor of any supported type                                |
+| `lhs_batching_dimensions`    | 1-dimensional tensor constant of type `si64`                |
+| `rhs_batching_dimensions`    | 1-dimensional tensor constant of type `si64`                |
+| `lhs_contracting_dimensions` | 1-dimensional tensor constant of type `si64`                |
+| `rhs_contracting_dimensions` | 1-dimensional tensor constant of type `si64`                |
+| `precision_config`           | variadic number of enum of `DEFAULT`, `HIGH`, and `HIGHEST` |
```

Comment:
I'm wondering whether we should use "or" here since it doesn't make sense for the two `precision_config` values to have different enum types (e.g. [DEFAULT, HIGHEST]) since this would result in opposing precision.
```
variadic number of enum of `DEFAULT`, `HIGH`, or `HIGHEST`
```

Reply:
Also noticed that `PACKED_NIBBLE` has been added recently:
https://github.com/tensorflow/tensorflow/blob/master/tensorflow/compiler/xla/xla_data.proto#L788

Reply:
Nice find! Yes, indeed it's been added not so long ago. We have https://github.com/openxla/stablehlo/issues/742 to track potentially adding that to StableHLO.